File: copilot-token-budget/extension/src/livebilling/refresher.ts
// extension/src/livebilling/refresher.ts — orchestrate cache-aware fetching of live billing data.
// Mirrors the Go refresher pattern for VS Code extension.

import { GitHubEntitlementFetcher } from './fetcher';
import {
  loadLiveBillingConfig,
  resolveLiveBillingAuth,
} from './config';
import {
  loadLiveBillingCache,
  saveLiveBillingCache,
  cacheIsFresh,
  newLiveBillingCacheEntry,
} from './cache';
import { LiveBillingSnapshot, LiveBillingCacheEntry } from '../types';

export interface RefreshResult {
  snapshot: LiveBillingSnapshot | undefined;
  error: string;
  cached: boolean;
}

export async function refreshLiveBilling(now: Date = new Date()): Promise<RefreshResult> {
  const config = loadLiveBillingConfig();
  const auth = resolveLiveBillingAuth(config);

  // If live billing is disabled, return nil immediately.
  if (auth.disabled) {
    return {
      snapshot: undefined,
      error: 'live billing disabled',
      cached: false,
    };
  }

  // If auth is not ready, return nil with error message.
  if (!auth.ready) {
    console.error(
      `copilot-budget: auth not ready (${auth.mode}); using estimated quota`
    );
    return {
      snapshot: undefined,
      error: auth.message,
      cached: false,
    };
  }

  // Try to load the existing cache.
  const cached = loadLiveBillingCache();
  if (cached && cacheIsFresh(cached, now)) {
    // Cache is fresh; return it without fetching.
    const snapshot = cached.snapshot;
    const h = hoursAgo(snapshot.lastRefreshedAt, now);
    snapshot.sourceLabel = `(authoritative, cached ~${h}h ago)`;
    return {
      snapshot,
      error: '',
      cached: true,
    };
  }

  // Cache is missing or stale. Fetch fresh data.
  console.error('copilot-budget: fetching live quota from GitHub...');

  const fetcher = new GitHubEntitlementFetcher(
    auth.config.gitHubAPIUrl || 'https://api.github.com',
    auth.token,
    auth.config.requestTimeoutSecs,
    auth.config.dryRun
  );

  let quota: number;
  try {
    quota = await fetcher.fetchEntitlements(config.orgSlug);
  } catch (err) {
    // Fetch failed. Log and return nil (graceful degradation).
    const errMsg = err instanceof Error ? err.message : String(err);
    if (errMsg.includes('timeout') || errMsg.includes('timed out')) {
      console.error('copilot-budget: GitHub API timed out; using estimated quota');
    } else if (errMsg.includes('401')) {
      console.error('copilot-budget: GitHub token invalid; using estimated quota');
    } else {
      console.error(
        `copilot-budget: fetch failed (${errMsg}); using estimated quota`
      );
    }
    return {
      snapshot: undefined,
      error: errMsg,
      cached: false,
    };
  }

  // Create a new snapshot with the fetched quota.
  const snapshot: LiveBillingSnapshot = {
    orgSlug: config.orgSlug,
    scope: 'org aggregate',
    sourceLabel: '(authoritative, live)',
    availability: 'available',
    lastRefreshedAt: now,
    asOf: now,
    credits: quota,
    error: undefined,
  };

  // Save to cache with TTL.
  const ttlHours = config.cacheMaxAgeHours;
  const cacheEntry = newLiveBillingCacheEntry(snapshot, undefined, ttlHours, now);
  try {
    saveLiveBillingCache(cacheEntry);
  } catch (err) {
    // Cache write failed, but we still have the fetched data, so log and continue.
    console.error(
      `copilot-budget: cannot save cache (${err instanceof Error ? err.message : err}); continuing without cache`
    );
  }

  console.error(`copilot-budget: fetched live quota ${quota} from GitHub`);

  return {
    snapshot,
    error: '',
    cached: false,
  };
}

// hoursAgo calculates the number of hours between time t and now.
// Returns at least 1 if t is less than 1 hour ago.
function hoursAgo(t: Date, now: Date): number {
  if (!t || t.getTime() === 0) {
    return 0;
  }
  const ms = now.getTime() - t.getTime();
  if (ms <= 0) {
    return 1;
  }
  return Math.max(1, Math.floor(ms / (60 * 60 * 1000)));
}

File: copilot-token-budget/core/internal/livebilling/refresher.go
// Package livebilling/refresher orchestrates fetching live billing data,
// checking cache freshness, updating cache, and returning snapshots.
package livebilling

import (
	"context"
	"fmt"
	"os"
	"time"
)

// RefreshSnapshot wraps the result of a refresh attempt.
type RefreshSnapshot struct {
	Snapshot *OrgBillingSnapshot
	Error    string
	Cached   bool
}

// Refresher orchestrates cache-aware fetching of live billing data.
type Refresher struct {
	cfg    Config
	auth   AuthResolution
	cache  CacheEntry
	cached bool
}

// NewRefresher constructs a Refresher from config and auth resolution.
func NewRefresher(cfg Config, auth AuthResolution) *Refresher {
	return &Refresher{
		cfg:  cfg,
		auth: auth,
	}
}

// Refresh attempts to fetch live billing data, respecting cache freshness.
// On success, it:
//   1. Checks if cached data is still fresh; if so, returns it.
//   2. Otherwise calls Fetcher.FetchEntitlements() and updates the cache.
//   3. Returns a snapshot with source label reflecting the outcome.
//
// On error (network, auth, or config), it logs to stderr and returns
// a nil snapshot (graceful degradation to estimated mode).
func (r *Refresher) Refresh(ctx context.Context) RefreshSnapshot {
	now := time.Now()

	// If live billing is disabled or not ready, return nil immediately.
	if r.auth.Disabled {
		return RefreshSnapshot{
			Snapshot: nil,
			Error:    "live billing disabled",
			Cached:   false,
		}
	}

	if !r.auth.Ready {
		fmt.Fprintf(os.Stderr, "livebilling: auth not ready (%s); using estimated quota\n", r.auth.Mode)
		return RefreshSnapshot{
			Snapshot: nil,
			Error:    r.auth.Message,
			Cached:   false,
		}
	}

	// Try to load the existing cache.
	cached, err := LoadCache()
	if err == nil && cached.IsFresh(now) {
		// Cache is fresh; return it without fetching.
		snapshot := cached.Snapshot
		h := hoursAgo(snapshot.LastRefreshedAt, now)
		snapshot.SourceLabel = fmt.Sprintf("(authoritative, cached ~%dh ago)", h)
		return RefreshSnapshot{
			Snapshot: &snapshot,
			Error:    "",
			Cached:   true,
		}
	}

	// Cache is missing or stale. Fetch fresh data.
	fmt.Fprintf(os.Stderr, "livebilling: fetching live quota from GitHub...\n")

	fetcher := NewFetcher(r.cfg, r.auth.Token)
	quota, fetchErr := fetcher.FetchEntitlements(ctx, r.cfg.OrgSlug)

	if fetchErr != nil {
		// Fetch failed. Log and return nil (graceful degradation).
		if _, ok := fetchErr.(interface{ Timeout() bool }); ok {
			fmt.Fprintf(os.Stderr, "livebilling: GitHub API timed out; using estimated quota\n")
		} else {
			switch fetchErr.Error() {
			case fmt.Sprintf("fetcher: GitHub API returned 401"):
				fmt.Fprintf(os.Stderr, "livebilling: GitHub token invalid; using estimated quota\n")
			default:
				fmt.Fprintf(os.Stderr, "livebilling: fetch failed (%v); using estimated quota\n", fetchErr)
			}
		}
		return RefreshSnapshot{
			Snapshot: nil,
			Error:    fetchErr.Error(),
			Cached:   false,
		}
	}

	// Create a new snapshot with the fetched quota.
	snapshot := OrgBillingSnapshot{
		OrgSlug:         r.cfg.OrgSlug,
		Scope:           ScopeOrgAggregate,
		SourceLabel:     "(authoritative, live)",
		Availability:    AvailabilityAvailable,
		LastRefreshedAt: now,
		AsOf:            now,
		Credits:         float64(quota),
		Error:           "",
	}

	// Save to cache with TTL.
	ttl := time.Duration(r.cfg.CacheMaxAgeHours) * time.Hour
	cacheEntry := NewCacheEntry(snapshot, nil, ttl, now)
	if err := SaveCache(cacheEntry); err != nil {
		// Cache write failed, but we still have the fetched data, so log and continue.
		fmt.Fprintf(os.Stderr, "livebilling: cannot save cache (%v); continuing without cache\n", err)
	}

	fmt.Fprintf(os.Stderr, "livebilling: fetched live quota %d from GitHub\n", quota)

	return RefreshSnapshot{
		Snapshot: &snapshot,
		Error:    "",
		Cached:   false,
	}
}
